pkg/webhook/v1alpha2/applicationconfiguration/helper.go

```diff
@@ -16,32 +15,102 @@ import (
+	workloadDefinition v1alpha2.WorkloadDefinition
+	workloadContent    unstructured.Unstructured
+	validatingTraits   []ValidatingTrait
+}
+
+// ValidatingTrait is used for validating Trait
+type ValidatingTrait struct {
+	componentTrait v1alpha2.ComponentTrait
+
+	// below data is convenient for validation
+	traitDefinition v1alpha2.TraitDefinition
+	traitContent    unstructured.Unstructured
+}
+
+// PrepareForValidation prepares data for validations to avoiding repetitive GET/unmarshal operations
+func (v *ValidatingAppConfig) PrepareForValidation(ctx context.Context, c client.Reader, dm discoverymapper.DiscoveryMapper, ac *v1alpha2.ApplicationConfiguration) []error {
+	var allErrors []error
+	v.appConfig = *ac
+	v.validatingComps = make([]ValidatingComponent, len(ac.Spec.Components))
+	for _, acc := range ac.Spec.Components {
+		tmp := ValidatingComponent{}
+		tmp.appConfigComponent = acc
+
+		if acc.ComponentName != "" {
+			tmp.compName = acc.ComponentName
+		} else {
+			tmp.compName = acc.RevisionName
+		}
+		comp, _, err := util.GetComponent(ctx, c, acc, ac.Namespace)
+		if err != nil {
+			allErrors = append(allErrors, errors.Wrapf(err,
+				errFmtGetComponent, tmp.compName))
+			continue
 		}
-		td, err := util.FetchTraitDefinition(ctx, client, dm, ut)
-		if err != nil && !apierrors.IsNotFound(err) {
-			return false, errors.Wrapf(err, errFmtGetTraitDefinition, ut.GetAPIVersion(), ut.GetKind(), ut.GetName())
+		tmp.component = *comp
+
+		// get worload content from raw
+		wl := unstructured.Unstructured{}
+		if err := json.Unmarshal(comp.Spec.Workload.Raw, &wl); err != nil {
+			allErrors = append(allErrors, errors.Wrapf(err,
+				errFmtUnmarshalWorkload, tmp.compName))
+		}
+		tmp.workloadContent = wl
+
+		// get workload definition
+		wlDef, err := util.FetchWorkloadDefinition(ctx, c, dm, &wl)
+		if err != nil {
+			allErrors = append(allErrors, errors.Wrapf(err,
+				errFmtGetWorkloadDefinition, tmp.compName))
 		}
-		if td.Spec.RevisionEnabled {
-			// if any traitDefinition's RevisionEnabled is true
-			// then the component is versioning enabled
-			return true, nil
+		tmp.workloadDefinition = *wlDef
+
+		tmp.validatingTraits = make([]ValidatingTrait, len(acc.Traits))
+		for _, t := range acc.Traits {
+			tmpT := ValidatingTrait{}
+			tmpT.componentTrait = t
+			// get trait content from raw
+			tContent := unstructured.Unstructured{}
+			if err := json.Unmarshal(t.Trait.Raw, &tContent); err != nil {
+				allErrors = append(allErrors, errors.Wrapf(err,
+					errFmtUnmarshalTrait, tmp.compName))
+			}
+			tmpT.traitContent = tContent
+
+			// get trait definition
+			tDef, err := util.FetchTraitDefinition(ctx, c, dm, &tContent)
+			if err != nil {
+				allErrors = append(allErrors, errors.Wrapf(err,
+					errFmtGetTraitDefinition, tmp.compName))
+			}
+			tmpT.traitDefinition = *tDef
+			tmp.validatingTraits = append(tmp.validatingTraits, tmpT)
 		}
 	}
-	return false, nil
+	// remove nil error
+	return allErrors
```

Comment:
just return nil, if no errors, allErrors cause confusion

Reply:
got it.